tokio/src/task/scope/intrusive_double_linked_list.rs

```diff
@@ -0,0 +1,628 @@
+//! An intrusive double linked list of data
+
+use core::marker::PhantomPinned;
+use core::ops::{Deref, DerefMut};
+use core::ptr::null_mut;
+
+/// A node which carries data of type `T` and is stored in an intrusive list
+#[derive(Debug)]
+pub(crate) struct ListNode<T> {
+    /// The previous node in the list. null if there is no previous node.
+    prev: *mut ListNode<T>,
+    /// The next node in the list. null if there is no previous node.
+    next: *mut ListNode<T>,
+    /// The data which is associated to this list item
+    data: T,
+    /// Prevents `ListNode`s from being `Unpin`. They may never be moved, since
+    /// the list semantics require addresses to be stable.
+    _pin: PhantomPinned,
+}
+
+impl<T> ListNode<T> {
+    /// Creates a new node with the associated data
+    pub(crate) fn new(data: T) -> ListNode<T> {
+        ListNode::<T> {
+            prev: null_mut(),
+            next: null_mut(),
+            data,
+            _pin: PhantomPinned,
+        }
+    }
+}
+
+impl<T> Deref for ListNode<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.data
+    }
+}
+
+impl<T> DerefMut for ListNode<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        &mut self.data
+    }
+}
+
+/// An intrusive linked list of nodes, where each node carries associated data
+/// of type `T`.
+#[derive(Debug)]
+pub(crate) struct LinkedList<T> {
+    head: *mut ListNode<T>,
+    tail: *mut ListNode<T>,
+}
+
+impl<T> LinkedList<T> {
+    /// Creates an empty linked list
+    pub(crate) fn new() -> Self {
+        LinkedList::<T> {
+            head: null_mut(),
+            tail: null_mut(),
+        }
+    }
+
+    /// Consumes the list and creates an iterator over the linked list.
+    /// This function is only safe as long as all pointers which are stored inside
+    /// the linked list are valid.
+    #[allow(dead_code)]
+    pub(crate) unsafe fn into_iter(self) -> LinkedListIterator<T> {
+        LinkedListIterator { current: self.head }
+    }
+
+    /// Consumes the list and creates an iterator over the linked list in reverse
+    /// direction.
+    /// This function is only safe as long as all pointers which are stored inside
+    /// the linked list are valid.
+    pub(crate) unsafe fn into_reverse_iter(self) -> LinkedListReverseIterator<T> {
+        LinkedListReverseIterator { current: self.tail }
+    }
+
+    /// Adds an item to the front of the linked list.
+    /// The function is only safe as long as valid pointers are stored inside
+    /// the linked list.
+    pub(crate) unsafe fn add_front(&mut self, item: *mut ListNode<T>) {
+        assert!(!item.is_null(), "Can not add null pointers");
```

Comment:
For example, if we took a `ptr::NonNull<ListNode<T>>` rather than`*mut ListNode<T>>` we could enforce this invariant in the type system, making it clearer to downstream code that calls into this...